datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -180,29 +251,103 @@ where
     ArrayHashSet { state, map }
 }
 
-/// Creates a `Box<dyn Set>` for the given list of `IN` expressions and `batch`
-fn make_set(array: &dyn Array) -> Result<Arc<dyn Set>> {
+/// Computes an [`ArrayHashSet`] for the provided [`StructArray`]
+fn make_struct_hash_set(array: &Arc<StructArray>) -> Result<ArrayHashSet> {
+    let state = RandomState::new();
+    let mut map: HashMap<usize, (), ()> =
+        HashMap::with_capacity_and_hasher(array.len(), ());
+
+    // Compute hashes for all rows
+    let mut row_hashes = vec![0u64; array.len()];
+    create_hashes_from_arrays(&[array.as_ref()], &state, &mut row_hashes)?;
+
+    // Build hash set, deduplicating based on struct equality
+    let insert_value = |idx: usize| {
+        let hash = row_hashes[idx];
+        if let RawEntryMut::Vacant(v) =
+            map.raw_entry_mut().from_hash(hash, |&existing_idx| {
+                // Compare the two struct rows for equality
+                // Use slice to get single-row arrays for comparison
+                let existing_row = array.slice(existing_idx, 1);
+                let current_row = array.slice(idx, 1);
+
+                // Use compare_op_for_nested for struct equality
+                match compare_with_eq(&existing_row, &current_row, true) {
+                    Ok(result) => result.value(0),
+                    Err(_) => false,
+                }
+            })
+        {
+            v.insert_with_hasher(hash, idx, (), |&x| row_hashes[x]);
+        }
+    };
+
+    match array.nulls() {
+        Some(nulls) => {
+            BitIndexIterator::new(nulls.validity(), nulls.offset(), nulls.len())
+                .for_each(insert_value)
+        }
+        None => (0..array.len()).for_each(insert_value),
+    }
+
+    Ok(ArrayHashSet { state, map })
+}
+
+/// Creates a `Box<dyn Set>` for the given list of `IN` expressions and `batch`.
+/// Accepts a parameter `f` which allows access to the underlying `ArrayHashSet`
+/// and is used by [`in_list_from_array`] to collect unique indices.
+fn make_set<F>(array: &dyn Array, mut f: F) -> Result<Arc<dyn Set>>
+where
+    F: FnMut(ArrayHashSet) -> Result<ArrayHashSet>,
```

Comment:
This change is to add a closure so we can cheaply recover the indexes of the unique items in the list re-using the work we did to build the hash table to build the equality check list.